app/uk/gov/hmrc/hec/controllers/IFController.scala

```diff
@@ -0,0 +1,123 @@
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class IFController @Inject() (
+  IFService: IFService,
+  cc: ControllerComponents
+)(implicit ec: ExecutionContext)
+    extends BackendController(cc)
+    with Logging {
+
+  private def correlationId = UUID.randomUUID().toString
+
+  /**
+    * Fetch individual user's self-assessment status
+    * @param utr Self-assessment UTR
+    * @param taxYear Start tax year
+    * @return Self-assessment status with UTR and tax year
+    */
+  def getSAStatus(utr: String, taxYear: String): Action[AnyContent] = Action.async { implicit request =>
+    val sautrValidation   = SAUTR.fromString(utr).toValidNel("Invalid SAUTR")
+    val taxYearValidation = TaxYear.fromString(taxYear).toValidNel("Invalid tax year")
+
+    @SuppressWarnings(Array("org.wartremover.warts.Any"))
+    val validation = (sautrValidation, taxYearValidation).mapN((_, _))
+    validation match {
+      case Valid((utr, year)) =>
+        IFService
+          .getSAStatus(utr, year, correlationId)
+          .fold(
+            {
+              case IFServiceImpl.DataError(msg)  =>
+                logger.warn(s"[correlationId:$correlationId] $msg")
+                NotFound
+              case IFServiceImpl.BackendError(e) =>
+                logger.warn(s"[correlationId:$correlationId] Could not fetch SA status", e)
+                InternalServerError
+            },
+            saStatus => Ok(Json.toJson(saStatus))
+          )
+
+      case Invalid(e) => Future.successful(BadRequest(e.toList.mkString("; ")))
+    }
+  }
+
+  /**
+    * Fetch a company's corporation tax status
+    * @param utr The corporation tax UTR
+    * @param startDate The start date of the tax period
+    * @param endDate The end date of the tax period
+    * @return Corporation tax status & accounting periods
+    */
+  def getCTStatus(
+    utr: String,
+    startDate: String,
+    endDate: String
+  ): Action[AnyContent] = Action.async { implicit request =>
+    def parsedDate(dateStr: String, error: String): Validated[NonEmptyList[String], LocalDate] =
+      try LocalDate.parse(dateStr, DateTimeFormatter.ISO_DATE).valid
+      catch {
+        case _: Exception => invalidNel(error)
+      }
+
+    val ctutrValidation: ValidatedNel[String, CTUTR]        = CTUTR.fromString(utr).toValidNel("Invalid CTUTR")
+    val fromDateValidation: ValidatedNel[String, LocalDate] = parsedDate(startDate, "Invalid startDate format")
+    val toDateValidation: ValidatedNel[String, LocalDate]   = parsedDate(endDate, "Invalid endDate format")
+
+    @SuppressWarnings(Array("org.wartremover.warts.Any"))
+    val validation = (ctutrValidation, fromDateValidation, toDateValidation).mapN((_, _, _))
+    validation match {
+      case Valid((utr, from, to)) =>
+        IFService
+          .getCTStatus(utr, from, to, correlationId)
+          .fold(
+            {
+              case IFServiceImpl.DataError(msg)  =>
+                logger.warn(msg)
```

Comment:
it'd be handy to have the correlation id in these messages like you've done above. Could we also get an info log just before the service is called which shows the correlation id just so we have more points to trace it. It's say something like `"Getting SA/CT status with correlation id $correlationId"`